vdr/ambassador.go

```diff
@@ -37,20 +49,243 @@ type Ambassador interface {
+	// This should succeed since the signature of the network document has already been verified.
+	pKey, err := n.keyResolver.GetPublicKey(document.SigningKeyID(), document.SigningTime())
+	if err != nil {
+		return fmt.Errorf("unable to resolve signingkey: %w", err)
+	}
+	signingKey, err := jwk.New(pKey)
+	if err != nil {
+		return fmt.Errorf("could not parse public key into jwk: %w", err)
+	}
+	// Create thumbprint
+	signingKeyThumbprint, err := signingKey.Thumbprint(thumbprintAlg)
+	if err != nil {
+		return fmt.Errorf("unable to generate network document signing key thumbprint: %w", err)
+	}
+
+	// Check if the signingKey is listed as a valid authenticationMethod in one of the controllers
+	keyToSign, err := n.findKeyByThumbprint(signingKeyThumbprint, controllerVerificationRelationships)
+	if err != nil {
+		return fmt.Errorf("unable to find signingKey by thumprint in controllers: %w",err)
+	}
+	if keyToSign == nil {
+		return fmt.Errorf("network document not signed by one of its controllers")
+	}
+
+	// TODO: perform all these tests:
+	// Take authenticationMethod keys from the controllers
+	// Check if network signingKeyID is one of authenticationMethods of the controller
+	//
+	// For each verificationMethod in the next version document
+	// 		check if the provided key thumbprint matches the corresponding thumbprint in the key store
+	// Take diff of verificationMethods between next and current versions:
+	// if new verificationMethod is added:
+	// 		Add public key to key store
+	// if verificationMethod is removed:
+	//		Mark keyID as expired since the updatedAt time from new DID document
+
+	// make a diff of the controllers
+	// 	if controller is added
+	//		check if it is known.
+	updatedAt := document.SigningTime()
+	documentMetadata := types.DocumentMetadata{
+		Created:    currentDIDMeta.Created,
+		Updated:    &updatedAt,
+		Version:    document.TimelineVersion(),
+		TimelineID: document.TimelineID(),
+		Hash:       document.PayloadHash(),
+	}
+	return n.didStore.Update(proposedDIDDocument.ID, currentDIDMeta.Hash, proposedDIDDocument, &documentMetadata)
+}
+
+// checkSubscriberDocumentIntegrity performs basic integrity checks on the SubscriberDocument fields:
+func checkSubscriberDocumentIntegrity(document dag.SubscriberDocument) error {
+	// check the payload type:
+	if document.PayloadType() != didDocumentType {
+		return fmt.Errorf("wrong payload type for this subscriber. Can handle: %s, got: %s", didDocumentType, document.PayloadType())
+	}
+
+	// PayloadHash must be set
+	if document.PayloadHash().Empty() {
+		return fmt.Errorf("payloadHash must be provider")
+	}
+
+	// Signing time should be set and lay in the past:
+	// allow for 2 seconds clock skew
+	if document.SigningTime().IsZero() || document.SigningTime().After(time.Now().Add(2*time.Second)) {
+		return fmt.Errorf("signingTime must be set and in the past")
+	}
+
+	if isUpdate(document) {
```

Comment:
double isUpdate, 1 from the callback and 1 from integrityCheck. This check could be splitted in 2 functions removing the need for this check.

Reply:
That would be 3 functios than: general checks, checks for create and for updates. I see that it is technacally correct, but I don't think 3 seperate functions are better that 1 with a redundant check.